### DISCUSSION TITLE
Markdown Preview

### DISCUSSION THREAD

**Discussion by txtyash** [helix-editor/helix#11325] [Ideas]
Markdown preview similar to this plugin:
https://github.com/davidgranstrom/nvim-markdown-preview

**Comment by CptPotato** [comment]
An in-terminal preview using the builtin markdown view would be a nice option, too.

**Comment by txtyash** [comment]
> An in-terminal preview using the builtin markdown view would be a nice option, too.

I don't think so. I've used similar with neovim and it's terrible. It has all wrong colors and no way to view images.

But I wouldn't say no to both types of previews.

**Comment by felix-u** [comment]
Subjectively, this sounds more like a plugin than a core feature.

**Comment by txtyash** [comment]
@felix-u Ya, a plugin  would suit better.

**Comment by Jedsek** [comment]
This should be a plugin

**Comment by aikomastboom** [comment]
I found this: https://pypi.org/project/markdown-server/

might be helpful

**Comment by MDeiml** [comment]
Markdown preview could actually (efficiently) be handled by a language server. There is already https://github.com/euclio/mdpls which does exactly that.

**Comment by MDeiml** [comment]
Actually installing that and adding
```toml
[[language]]
name = "markdown"
language-server = { command = "/path/to/mdpls" }
config = { markdown.preview.auto = true, markdown.preview.browser = "firefox" }
```
to your `language.toml` already gives you a markdown preview. The repo I linked seems to be missing a few core features though, e.g. images don't work and I would love some auto-scrolling.

**Comment by TheCaptainCraken** [comment]
@MDeiml I tried that but it looks like it does not support the $\neq$ :(

**Comment by MDeiml** [comment]
Not sure if latex is supported for that repo, seems like it uses https://github.com/euclio/aurelius under the hood which might support latex? I'm not sure.

You could open an issue in that repo and ask the author.

**Comment by TheCaptainCraken** [comment]
Yep I looked into that and it should support latex. I checked and it supports all other symbols like $\sqrt x$ $\gt$ etc.
It actually prints the $\ne$ butit renders it like \  = wich is odd. I'll open an issue there.

**Comment by 0xlilith** [comment]
> Actually installing that and adding
> 
> ```toml
> [[language]]
> name = "markdown"
> language-server = { command = "/path/to/mdpls" }
> config = { markdown.preview.auto = true, markdown.preview.browser = "firefox" }
> ```
> 
> to your `language.toml` already gives you a markdown preview. The repo I linked seems to be missing a few core features though, e.g. images don't work and I would love some auto-scrolling.

how do I open the preview?

**Comment by thorstenweber83** [comment]
@0xlilith  
`:lsp-workspace-command Open\ Preview` should do the trick!
If it does not open, mdpls probably can not find the browser you specified inside your path.

**Comment by MDeiml** [comment]
The preview should also open automatically if you open a markdown file. It works for me at least.

**Comment by sgon00** [comment]
Hopefully, helix will have better markdown preview or highlight support in the future. So far, I am using vscode for markdown editing.

**Comment by jem-os** [comment]
> Actually installing that and adding
> 
> ```toml
> [[language]]
> name = "markdown"
> language-server = { command = "/path/to/mdpls" }
> config = { markdown.preview.auto = true, markdown.preview.browser = "firefox" }
> ```
> 
> to your `language.toml` already gives you a markdown preview. The repo I linked seems to be missing a few core features though, e.g. images don't work and I would love some auto-scrolling.

Is there a way to control the presentation theme in the Firefox preview? e.g. not white text on black background. Don't think I want to manage the css directly in the markdown files as that would impact it's presentation everywhere

**Comment by MDeiml** [comment]
Doesn't seem like it, but it doesn't seem to hard to implement yourself. Maybe create an issue or pull request in https://github.com/euclio/mdpls? Or you could just fork that repo and edit the template to your liking.

**Comment by David-Else** [comment]
I am using `gh extension install yusukebe/gh-markdown-preview`, it is great and has live reload, but no mirroring the line position in Helix and the browser. Can anything do that in Helix? Does [mdpls](https://github.com/euclio/mdpls) do anything extra?

**Comment by arcstur** [comment]
> Actually installing that and adding
> 
> ```toml
> [[language]]
> name = "markdown"
> language-server = { command = "/path/to/mdpls" }
> config = { markdown.preview.auto = true, markdown.preview.browser = "firefox" }
> ```
> 
> to your `language.toml` already gives you a markdown preview. The repo I linked seems to be missing a few core features though, e.g. images don't work and I would love some auto-scrolling.

This is awesome, thank you very much. This, however, disables `marksman`, which has cool features (creating Table of Contents, [[wiki lins]], [[#heading links]], etc). So I instead use this:
```toml
[[language]]
name = "markdown"
language-servers = ["marksman", "mdpls"]

[language-server.mdpls]
command = "/home/arcstur/.cargo/bin/mdpls"
config = { markdown.preview.auto = true }
```

Open Preview doesn't appear in the workspace commands, but it always open a new preview when opening a markdown file. I don't know how to merge both workspace commands of the two LSPs.

**Reply by rahulTPatil** [reply]
> > Actually installing that and adding
> > ```toml
> > [[language]]
> > name = "markdown"
> > language-server = { command = "/path/to/mdpls" }
> > config = { markdown.preview.auto = true, markdown.preview.browser = "firefox" }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to your `language.toml` already gives you a markdown preview. The repo I linked seems to be missing a few core features though, e.g. images don't work and I would love some auto-scrolling.
> 
> This is awesome, thank you very much. This, however, disables `marksman`, which has cool features (creating Table of Contents, [[wiki lins]], [[#heading links]], etc). So I instead use this:
> 
> ```toml
> [[language]]
> name = "markdown"
> language-servers = ["marksman", "mdpls"]
> 
> [language-server.mdpls]
> command = "/home/arcstur/.cargo/bin/mdpls"
> config = { markdown.preview.auto = true }
> ```
> 
> Open Preview doesn't appear in the workspace commands, but it always open a new preview when opening a markdown file. I don't know how to merge both workspace commands of the two LSPs.

I was able to get goto-defnition and Open Preview working (as I had different notes linked) with these 2 working, just have to switch the order of LSPs
`language-servers = ["mdpls", "marksman"]`

**Comment by superiums** [comment]
modern browser is so weight for preview task.
maybe something light more suit for this.
i'll tried dillo , but not work.

vimb, midori works. 
any more good one?

**Comment by arcstur** [comment]
I think that choosing a good theme makes a lot of difference. I saw that only a few themes have different colors for heading levels. Kanagawa, sonokai and papercolor have, for example. Only a few themes use colors for italic and bold, monokai being one of them. I think that using that, markdown already looks a lot better.

The only theme I found that had different colors _both_ for h1/h2 and for bold/italic is onelight. I wonder if we can modify other themes to also differentiate colors on bold/italic/h1/h2. Doing that I think will make the markdown experience much better.

Markdown is made such that seeing it as code already makes it good, so I don't think we need different sizes for headings for example.

The only thing I would miss is having image previews. If someone that understands the codebase could guide me into where to start developing this, I would like it very much. I know it would be a plugin and we still need a plugin system haha.

**Comment by arcstur** [comment]
Looking at the code in /runtime/themes/ we can see that some themes define what I mentioned.

Kanagawa using colors for heading levels:

```toml kanagawa
"markup.heading.marker" = "springViolet2"
"markup.heading.1" = { fg = "surimiOrange", modifiers = ["bold"] }
"markup.heading.2" = { fg = "carpYellow", modifiers = ["bold"] }
"markup.heading.3" = { fg = "waveAqua2", modifiers = ["bold"] }
"markup.heading.4" = { fg = "lightBlue", modifiers = ["bold"] }
"markup.heading.5" = { fg = "oniViolet", modifiers = ["bold"] }
"markup.heading.6" = { fg = "springViolet1", modifiers = ["bold"] }
```

Monokai using colors for bold and italic

```toml
"markup.bold" = { fg = "yellow", modifiers = ["bold"] }
"markup.italic" = { fg = "magenta", modifiers = ["italic"] }
```

In my personal opinion, having colors for both heading levels and bold + italic make reading markdown much easier.

I wonder if we also can do that for other themes, or that is defined by the theme authors and we would violate their intetions. What are the maintainer thoughts on this?

**Comment by mzabani** [comment]
I wrote a bash script that previews anything pandoc can render in the terminal itself, and detects modifications to the file to reload. My helix-config-foo isn't great so I'm not sure if this can be integrated well; I'm currently using it in a separate terminal tab/window.
I hope this is useful to someone! It requires inotify-tools, [termpdf.py](https://github.com/dsanson/termpdf.py) or some other in-terminal PDF reader (do change the code if you're using something else) and pandoc.

````bash
#!/usr/bin/env bash

set -eo pipefail

FILE="$(realpath "$1")"
TMPDIR="$(mktemp -d)"
PDF="$TMPDIR/rendered.pdf"
trap "pkill -P $$ && rm -rf \"$TMPDIR\"" exit

regen() {
	inotifywait -q -e modify "$FILE" "$TMPDIR/closing.signal" > "$TMPDIR/inotify.log"
	grep -q "$TMPDIR/closing.signal" "$TMPDIR/inotify.log" || {
		pandoc "$FILE" -o "$PDF"
		pkill -P "$$" # Kills children of these script, which can only be termpdf.py at this point
	}
}

pushd "$TMPDIR" >/dev/null # termpdf.py's log file will go here and will be deleted on exit
pandoc "$FILE" -o "$PDF"
touch closing.signal

while :
do
	regen &
	termpdf.py "$PDF" && popd>/dev/null && echo 1 >"$TMPDIR/closing.signal" && exit 0 # If the user quits the viewer, we can exit nicely
done
````

**Comment by TheXnlyGhxst** [comment]
I'm not sure how to do it, but this seems like a fairly simple solution if it could be built into Helix, may also be able to simply use it as a language server in languages.toml: https://github.com/jannis-baum/vivify

> Editor Support 
Vivify has a simple API to integrate your favorite editor so the viewer live updates to any changes as you are typing and the scrolling is smoothly synchronized! 
See below for a list of existing editor plugins. In case your favorite editor is not yet supported, use these as an example to write your own and add it to the list!

**Comment by pascalkuthe** [comment]
> Vivify has a simple API to integrate your favorite editor

... if your "favorite editor" is written in javascript. This is something that can be covered by an LSP but it does not fit into core.

**Reply by TheXnlyGhxst** [reply]
Good point, it would need to be entirely rewritten in that case.

**Comment by TheXnlyGhxst** [comment]
For anyone still looking for a good solution to this, I've found that [Glow](https://github.com/charmbracelet/glow) works pretty well, if you don't mind a non rust program. Now I just want to figure out how to add a custom command into helix so I can launch it within the editor.

Edit: Discovered [Inlyne](https://github.com/Inlyne-Project/inlyne) after typing this: made with rust, actively maintained, does HTML viewing, and updates live. Has less features than Glow but I think this will be my solution for now.

**Reply by adriangalilea** [reply]
> Now I just want to figure out how to add a custom command into helix so I can launch it within the editor.

You can use Zellij or TMUX and have a shortcut that opens up a popup/side-panel with the content, or as you mention later, you can have a program in another window that autoupdates. But again, I think Zellij would be best.

You can take a look at a yazi file-picker example in [this discussion](https://github.com/luccahuguet/zellij/issues/8#issuecomment-2247063920) all my dotfiles are public so you can take a look at them and get an idea. I've successfully implemented it to open up lazygit, taskwarrior-tui, yazi and [elia](https://github.com/darrenburns/elia) as a floating window through a helix shortcut.

In fact there's a "hack" that let's you open something like lazygit inside a helix buffer, but I'm not sure I can recommend that.
`C-g = [":new", ":insert-output lazygit", ":buffer-close!", ":redraw"]`

lazygit was the only program that I was able to open like that, but maybe you can try glow or inlyne too.

**Reply by FelixSchausberger** [reply]
@Gh0stGames did you ever figure out how to add a shortcut for glow? I can get a preview using the built in shell (like `:sh glow "$0"`) but I'm unable to use it with a shortcut.

**Reply by andreacfromtheapp** [reply]
thank you @TheXnlyGhxst for Inlyne! It is amazing :)

**Reply by andar1an** [reply]
Ive been hsing mdpls, excited try inlyne too!

**Reply by andar1an** [reply]
Does anyone know if it is possible to reference current file in language server configuration command? 

I have tried using ' inlyne view %buffer_name' but get an error about non-canonicalized path. 

With mdpls there is an auto preview config option, trying to get similar functionality.

**Reply by tdHiru** [reply]
I already posted this in another thread, but I managed to achieve something similar using a tool called **Leaf**.

   https://github.com/helix-editor/helix/discussions/11325#discussioncomment-17714137

That said, if anyone has successfully set this up using **glow** or **inlyne**, I would love to see how you did it. Please share your configuration!

**Reply by andar1an** [reply]
Thank you for sharing this, love tuis and would rather stay in terminal! I know I will be trying today haha

**Comment by David-Else** [comment]
I have some different solutions to preview Markdown in the browser: https://helix-editor-tutorials.com/tutorials/writing-documentation-and-prose-in-markdown-using-helix/#markdown-preview-in-the-browser

**Comment by nik-rev** [comment]
Helix already includes a markdown renderer, but it is an internal component that is used for rendering LSP documentation in hover menus

One could potentially re-use it when creating this feature

**Comment by taqtiqa-mark** [comment]
It appears that `mpls` is the maintained successor to `mdpls`,,,, https://github.com/mhersson/mpls

**Reply by andreacfromtheapp** [reply]
this is also amazing! (if one likes the browser better). I'm keeping both inlyne and mpls options.

**Comment by eureka-cpu** [comment]
> Helix already includes a markdown renderer, but it is an internal component that is used for rendering LSP documentation in hover menus
> 
> One could potentially re-use it when creating this feature

I messed around with this. I think the hard part is that things are done very rigidly in terms of when and how things get rendered. I have a branch where there is some extra command line options so you can do `:vsplit-preview` and it will keep you in the same node instead of switching to the new one. When I looked around for options in terms of rendering, there is `render_document` and `render_text`, but these functions don't take into account that you may want to have special case rendering. I'm still working on it, perhaps after staring at it long enough I'll come up with an idea of how to get un-editable text to render while still being able to focus the node, in a way that is reasonable.

In the mean time, I came up with a small solution, which just always renders markdown files on hover if the lsp does not have that capability so that it doesn't interfere with lsp's that do, like https://github.com/matkrin/md-lsp.

```diff
diff --git a/helix-term/src/commands/lsp.rs b/helix-term/src/commands/lsp.rs
index 0494db3e..a7a60aae 100644
--- a/helix-term/src/commands/lsp.rs
+++ b/helix-term/src/commands/lsp.rs
@@ -1042,6 +1042,33 @@ pub fn hover(cx: &mut Context) {
         .count()
         == 0
     {
+        // Always render markdown on hover if markdown language server does not have hover capabilities.
+        if doc.language_name().is_some_and(|lang| lang == "markdown") {
+            cx.jobs.callback({
+                let content = doc.text().to_string();
+                async move {
+                    let hovers = vec![(
+                        "helix-markdown-hover".into(),
+                        lsp::Hover {
+                            contents: lsp::HoverContents::Markup(lsp::MarkupContent {
+                                kind: lsp::MarkupKind::Markdown,
+                                value: content,
+                            }),
+                            range: None,
+                        },
+                    )];
+
+                    let call = move |editor: &mut Editor, compositor: &mut Compositor| {
+                        let contents = Hover::new(hovers, editor.syn_loader.clone());
+                        let popup = Popup::new(Hover::ID, contents).auto_close(true);
+                        compositor.replace_or_push(Hover::ID, popup);
+                    };
+                    Ok(Callback::EditorCompositor(Box::new(call)))
+                }
+            });
+            return;
+        }
+
         cx.editor
             .set_error("No configured language server supports hover");
         return;
```

Maybe this will help anyone else who is wanting to try it. Another thought I had is, instead of preview hover, or preview split, it may be nice to just have a markdown-render command that just renders the markdown on the page temporarily until something resets it, similarly to hover. This would be much easier than a split preview, since it would be very close to what the diff I posted above does. One could potentially create a new UI element called `Preview` instead of `Hover` and then essentially do this.

**Reply by eureka-cpu** [reply]
One other issue with the word preview though, is that it implies what most people would expect from a previewer, which would include things like image loading. I'm not sure if the project would be okay with adding something like that, though there's ample projects which can be used as reference for image rendering. As a fallback, it could just _not_ render images lol.

**Comment by sytherax** [comment]
My solution using `mcat`. This will open a full viewport preview of markdown or any file mcat supports. And press q to quit when done.

## mcat-preview.sh
```bash
#!/usr/bin/env bash
printf '\033[?1049h' > /dev/tty
mcat --paging always "$1" > /dev/tty
printf '\033[?1049l' > /dev/tty
```

## helix config
```toml
[keys.normal."\\"]
p = [":insert-output mcat-preview.sh \"%{buffer_name}\" > /dev/null", ":redraw"]
```

**Comment by tdHiru** [comment]
I found a neat solution for viewing Markdown files directly from Helix using a tool called **Leaf** (a terminal markdown viewer). 

Here is the configuration you can add to your `config.toml` to open the current buffer in Leaf:

```toml
[keys.normal."space"]
m = [':insert-output leaf "$(echo %{buffer_name} | sed "s|^~|$HOME|")" >/dev/tty', ":redraw"]
```

The `sed` command ensures that the `~` path expands to `$HOME` properly.

* Links: [Leaf Website](https://leaf.rivolink.mg/) / [GitHub Repository](https://github.com/rivolink/leaf)

I hope this helps anyone looking for a way to view Markdown files!

**Reply by andar1an** [reply]
As another tui option, or for inspo, there is also [glow](https://github.com/charmbracelet/glow)
